azure/cargo-check.yml

```diff
@@ -8,12 +8,12 @@ jobs:
 - job: ${{ parameters.name }}
   ${{ if eq(parameters.rust, 'stable') }}:
     ${{ if eq(parameters.err_on_warn, 'true') }}:
-      displayName: "env RUSTFLAGS=-Dwarnings cargo check"
+      displayName: "cargo check -Dwarnings"
     ${{ if ne(parameters.err_on_warn, 'true') }}:
       displayName: cargo check
   ${{ if ne(parameters.rust, 'stable') }}:
     ${{ if eq(parameters.err_on_warn, 'true') }}:
-      displayName: "env RUSTFLAGS=-Dwarnings cargo +${{ parameters.rust }} check"
+      displayName: "cargo +${{ parameters.rust }} check -Dwarnings"
```

Comment:
Should we allow `deprecated`?  It'll make things more stable.

Reply:
This is already `allow_fail`, so I'd be inclined to still give deprecation warnings. They're things you should probably fix, just like other warnings.

Reply:
1. This template does not have an `allow_fail`
2. This assumes that the caller will always pass `allow_fail` with `err_on_warn`.  In my use case, I do not want to be doing that.

Reply:
Ah, good catch! Fixed 1 in latest commit.

As for 2, it doesn't really assume that. If you specify `err_on_warn` without `allow_fail`, you are specifically saying "I want to treat warnings as CI errors". Seems reasonable for that to be a valid option?

Reply:
While so far I've gotten away with `deprecated` being enabled, you've brought up the concern over dependencies breaking people.  If we want to take that seriously and support people who want a pinned warnings check, then we'd need to disable that warning.

Reply:
My thinking here was that either you care about warnings, or you do not (this is `err_on_warn`). If you do, you either want them to cause red CI or you want them to cause yellow CI (this is `allow_fail`). We could add a third branch point which is "care about all but warnings caused by changes in dependencies" I suppose (what would it be called)? My guess is that that option is less common, but I have limited data on that.

Reply:
My care abouts
- Warnings make CI red
- Transient state (like dependencies deprecating things when you don't have a Cargo.lock checked in) should not make my CI red
  - Personally, I've never had a dependency make something deprecated so I've not worried about that state before.  This is more in theoretically possible land
- I should be able to maintain my MSRV ie deprecation warnings for Rust can only be emitted by the MSRV or else I can't get a "non-red" to approve my submissions.

As long as those conditions are met, I'm fine.

Reply:
Okay, so, I _think_ this satisfies all of those. Specifically, the default `stages.yml` will never make your CI red from warnings, it will only ever make it yellow. You _can_ choose to directly depend on `cargo-check.yml`, and then you can choose whether you want warnings to cause red CI (`err_on_warn: true` + `allow_fail: false`), yellow CI (`err_on_warn: true` + `allow_fail: true`), or green CI (`err_on_warn: false` + `allow_fail: false`; the default).

Thus, anything that'd make your CI red would be by choice in setting up your pipeline.

Now, what this PR does _not_ give you the ability to do is to run CI and have it _fail_ on warnings that are _not_ due to dependencies. I think that's specifically something you would like, but I also think that's something that's better handled in a new PR where we can explicitly figure out which warnings that would require allowing.

So, with all that said, my inclination would be to merge and then open an issue tracking "red CI from non-dependency warnings".